### ISSUE TITLE
Downstream device can't access network when upstream Ethernet Gadget using Exit Node

### ISSUE THREAD

**Issue by soulless-viewer** [tailscale/tailscale#13522]
### What is the issue?

I have the following setup:
- _[Raspberry]_ **Raspberry Pi Zero 2 W on Linux 6.6.20** _(Tailscale 1.74.0)_ - configured as an Ethernet Gadget providing network access (connected to Wi-Fi);
- _[MacBook]_ **MacBook with macOS Sonoma 14.2.1** _(no Tailscale)_ - connected to the Raspberry via USB-C, using it as a network provider;
- _[Exit Node 1]_ **Android 9** _(Tailscale 1.72.0)_ - set up as an exit node;
- _[Exit Node 2]_ Another **MacBook with macOS Sonoma 14.2.1** _(Tailscale 1.74.0)_ - set up as an alternative exit node just in case.

My goal is actually simple - I want MacBook's traffic to be routed through the Tailnet's Exit Node without installing Tailscale on it and with minimum configuration of macOS _(connect Ethernet Gadget and rock)_.

---
### Description

The setup works fine - the MacBook is connected to the Raspberry via USB-C, it uses `RNDIS/Ethernet Gadget` interface to communicate and the Raspberry is connected to Wi-Fi. Tailscale running on Raspberry is not causing any issues **as long as an exit node is not used**. 

However, when either Exit Node 1 or 2 is used, the MacBook loses internet access and can only communicate with the Raspberry Pi. All outgoing connections from the MacBook appear to stuck. Although, Tailscale itself is working fine, I'm confused as I expected Raspberry to route **all the traffic** to Exit Node including the traffic from downstream clients.

### Debugging

Using `tcpdump`, I confirmed that the MacBook’s requests are reaching the `tailscale0` interface on the Raspberry Pi, but they never reach their external destination when an exit node is in use.

```sh
[user@macbook ~]$ nc -v 1.1.1.1 80
<nothing here>
```
```sh
[user@raspberry ~]$ sudo tcpdump -i tailscale0 src 192.168.137.79 and dst 1.1.1.1
23:33:12.402085 IP 192.168.137.79.55902 > one.one.one.one.http: Flags [S], seq 2583318242, win 65535, length 0
...
```

I also checked `ip route`, `table 52` and all this related stuff _(too much info to share from the beginning)_ and everything seems to be working fine, but it's not...

Additionally, I tested with my own temp public endpoint and confirmed that requests do not reach the endpoint when an exit node is active so it's not about blocking responses or so.

### Steps to reproduce

A bit exotic, but:
1. Run Tailscale on Android or macOS as an Exit Node
2. Connect macOS client (without Tailscale installed) to the Ethernet Gadget with Linux and Tailscale running
3. Set Tailscale on Ethernet Gadget to use any Exit Node
4. macOS client is now cannot reach Internet

### OS

Linux, macOS, Android

### OS version

Raspbian (Linux 6.6.20), macOS Sonoma 14.2.1, Android 9

### Tailscale version

1.74.0, 1.74.0, 1.72.0

**Comment by agottardo** [comment]
Totally unrelated, but is there a Tailscale-related reason why you're still on macOS 14.2.1? You're missing out on 10 months worth of security updates (macOS 14.7 is the latest security patch for Sonoma).

**Comment by soulless-viewer** [comment]
> Totally unrelated, but is there a Tailscale-related reason why you're still on macOS 14.2.1? You're missing out on 10 months worth of security updates (macOS 14.7 is the latest security patch for Sonoma).

No, there is not. I'll upgrade it, but as you said - I don't think it might be related

**Comment by raggi** [comment]
You will need `--exit-node-allow-lan-access=true` (CLI) or "Allow local network access" (GUI) on the RPI so that LAN traffic is not blocked (to/from the Macbook). After that you'll also need to install your own NAT rules so that the RPI is acting as a NAT from the gadget interface to the tailscale interface.

**Comment by soulless-viewer** [comment]
Thanks @raggi,

I already tried that, but I either configured RPI's NAT rules incorrectly or this is not the case.
Could you please suggest how to configure NAT in this case? On the RPI I have 3 interfaces:
- wlan0 - net card / Wi-Fi
- usb0 - RNDIS related
- tailscale0 - Tailscale's virtual one

**Comment by soulless-viewer** [comment]
TL;DR

The issue is that the MacBook's traffic, which is forwarded through the Raspberry where Tailscale is running, is not marked. As a result, it is not masqueraded and gets dropped.

---

Analyzing `iptables` rules and routing with `tcpdump` and `iptables` logging I found out that the traffic is actually forwarded from `usb0` to `tailscale0`, but for some reason it cannot reach the destination.

I noticed that in the `POSTROUTING` chain of the `nat` table, a mark (`0x40000/0xff0000`) is being checked for matching. I found that the same mark is set when traffic is forwarded from the `tailscale0` interface. Putting two and two together, I realized this marking is used to identify Tailscale traffic after it leaves the `tailscale0` interface.

**So, I simply added the same marking to the traffic from `usb0`. I inserted a new `MARK` rule in the `ts-forward` chain of the `filter` table** _(the marking can be actually applied at any stage before masquerading)_.

```sh
$ sudo iptables -t filter -L FORWARD -n -v
Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
19769 5671K ts-forward  0    --  *      *       0.0.0.0/0            0.0.0.0/0           
 1492  669K ACCEPT     0    --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED

$ sudo iptables -t filter -L ts-forward -n -v
Chain ts-forward (1 references)
 pkts bytes target     prot opt in     out     source               destination         
 5465 1140K MARK       0    --  usb0   *       0.0.0.0/0            0.0.0.0/0            MARK xset 0x40000/0xff0000
 7252 3040K MARK       0    --  tailscale0 *       0.0.0.0/0            0.0.0.0/0            MARK xset 0x40000/0xff0000
16190 4735K ACCEPT     0    --  *      *       0.0.0.0/0            0.0.0.0/0            mark match 0x40000/0xff0000
    0     0 DROP       0    --  *      tailscale0  100.64.0.0/10        0.0.0.0/0           
 2087  267K ACCEPT     0    --  *      tailscale0  0.0.0.0/0            0.0.0.0/0

$ sudo iptables -t nat -L POSTROUTING -n -v
Chain POSTROUTING (policy ACCEPT 1635 packets, 136K bytes)
 pkts bytes target     prot opt in     out     source               destination         
 1761  164K ts-postrouting  0    --  *      *       0.0.0.0/0            0.0.0.0/0           

$ sudo iptables -t nat -L ts-postrouting -n -v
Chain ts-postrouting (1 references)
 pkts bytes target     prot opt in     out     source               destination         
  217 35090 MASQUERADE  0    --  *      *       0.0.0.0/0            0.0.0.0/0            mark match 0x40000/0xff0000
```

Since I don't have default rules for masquerading, the MacBook's traffic will only be completely forwarded through the `tailscale0` interface (when Tailscale is up, regardless of Exit Node usage). 

---

The only remaining task is to enable the marking only after checking whether or not an Exit Node is in use. This ensures that the traffic is routed only through the Exit Node. Right now I'm just using a basic Bash script to parse the `status` and `exit-node` commands output _(which is kinda 🩼)_.

**Comment by mdjhnson** [comment]
@soulless-viewer, from your last comment, it seems like you solved this but I'm not quite sure how to do it myself. Could you explain a little more, please?

**Comment by soulless-viewer** [comment]
@mdjhnson, as I wrote, for my specific I just added a marking rule in the `ts-forward` chain of the `filter` table to mark the traffic from the `usb0` interface allowing it to be masqueraded as it the same as the traffic from ‘tailscale0` interface.

If you have any specific questions I’d be happy to help you.

**Comment by mdjhnson** [comment]
Thanks. Basically, that's new to me. I haven't done anything with iptables before. Would you be able to explain how to add a marking rule for this case or is there something you recommend reading?

I'm also a little confused because I have a raspberry pi 5 and a raspberry pi zero 2 w. I'm able to access the internet when using the RP5 but not the RPzero2 when using it as an exit node. I've checked my ACL and even have tried allowing everything in the ACL.

When I run `sudo iptables -t filter -L ts-forward -n -v` on both devices, I get similar results. They both have a denied entry without marks.


## Raspberry Pi Zero 2 W
It is running a Raspberry Pi OS Lite (64-bit). It's a port of Debian Bookworm with no desktop environment.
```
matthewjohnson@johnsonremotepi:~ $ sudo iptables -t filter -L ts-forward -n -v
Chain ts-forward (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 MARK       0    --  tailscale0 *       0.0.0.0/0            0.0.0.0/0            MARK xset 0x40000/0xff0000
    0     0 ACCEPT     0    --  *      *       0.0.0.0/0            0.0.0.0/0            mark match 0x40000/0xff0000
    0     0 DROP       0    --  *      tailscale0  100.64.0.0/10        0.0.0.0/0           
    0     0 ACCEPT     0    --  *      tailscale0  0.0.0.0/0            0.0.0.0/0  
```

## Raspberry Pi 5
I'm pretty sure it is running Raspberry Pi OS (64-bit). It's a port of Debian Bookworm with the Raspberry Pi Desktop. I'm almost positive this is some version of the Raspberry Pi OS that HAS a desktop.
```
matthewjohnson@vpi5:~ $ sudo iptables -t filter -L ts-forward -n -v
Chain ts-forward (1 references)
 pkts bytes target     prot opt in     out     source               destination         
96828   28M MARK       0    --  tailscale0 *       0.0.0.0/0            0.0.0.0/0            MARK xset 0x40000/0xff0000
96828   28M ACCEPT     0    --  *      *       0.0.0.0/0            0.0.0.0/0            mark match 0x40000/0xff0000
    0     0 DROP       0    --  *      tailscale0  100.64.0.0/10        0.0.0.0/0           
 127K  103M ACCEPT     0    --  *      tailscale0  0.0.0.0/0            0.0.0.0/0
```

> [!IMPORTANT]
**EDIT:**
> I actually managed to fix my issue by [enabling IP forwarding](https://tailscale.com/kb/1019/subnets?tab=linux#enable-ip-forwarding). The reason I had it set up on the Raspberry Pi 5 and not the Zero 2 W was because I had subnet routing set up on the Pi 5.

**Comment by soulless-viewer** [comment]
I’m glad to see you managed to solve this. 

As you found out, described configuration related to the traffic marking is not applicable to your case since you have Tailscale installed on both devices.

But answering your question for historical purposes, to add a marking rule we can use the following command:

```sh
iptables -I ts-forward -i usb0 -j MARK —set-xmark 0x40000/0xff0000
```

- - -

> When I run `sudo iptables -t filter -L ts-forward -n -v` on both devices, I get similar results. They both have a denied entry without marks.

In order not to confuse other people I’d say «No, they don’t» :)

The `ts-forward` chain does not have a REJECT or DROP rule for unmarked traffic from what you shared. It’s possible that your `FORWARD` chain in the same table could contain such a rule, but who knows. As long as the system allows it (e.g., `net.ipv4.ip_forward=1`), the traffic should be forwarded.
